checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -362,57 +398,128 @@ private static int getGroupCount(Pattern p) {
-                group += 1;
-                if (i != 0 && regexp.charAt(i - 1) == '|') {
-                    nonNullGroups.remove((Integer) group);
+            } else if (regexp.charAt(i) == ')') {
+                if (!escaped && insideList == 0 && !quoted) { // ending of a construct.
+                    boolean closesGroup = openingWasGroup.pop();
+                    if (closesGroup) {
+                        int value = openingIndices.pop();
+                        if (i != regexp.length() - 1
+                                && "?|*".contains(Character.toString(regexp.charAt(i + 1)))) {
+                            nonNullGroups.remove(Integer.valueOf(value));
+                        } else if (i < length - 2
+                                && regexp.charAt(i + 1) == '{'
+                                && regexp.charAt(i + 2) == '0') {
+                            nonNullGroups.remove(Integer.valueOf(value));
+                        }
+                    }
+                } else if (escaped) {
+                    escaped = false;
                 }
-                openingIndices.push(group);
-            } else if (!escaped && !squareBracketOpen && regexp.charAt(i) == ')') {
-                if (notCapturing) {
-                    notCapturing = false;
-                    continue;
+            } else if (regexp.charAt(i) == '\\') {
+                if (!quoted) {
+                    escaped =
+                            !escaped; // if it is already escaped, '\' will be matched, if not, now
+                    // escaped becomes true.
+                } else if (i != length - 1 && regexp.charAt(i + 1) == 'E') {
+                    escaped = true;
                 }
-                int popped = openingIndices.pop();
-                if (i != regexp.length() - 1
-                        && "*?|".contains(Character.toString(regexp.charAt(i + 1)))) {
-                    nonNullGroups.remove((Integer) popped);
+            } else if (regexp.charAt(i) == '[') {
+                if (!escaped && !quoted) {
+                    insideList += 1;
+                    if (regexp.charAt(++i) == '\\') escaped = true;
+                } else if (escaped) {
+                    escaped = false;
                 }
-            } else if (!escaped && !squareBracketOpen && regexp.charAt(i) == '[') {
-                squareBracketOpen = true;
-            } else if (squareBracketOpen && regexp.charAt(i) == ']') {
-                squareBracketOpen = false;
-            }
-            if (!escaped && regexp.charAt(i) == '\\') {
-                escaped = true;
-            } else if (escaped) {
-                escaped = false;
+            } else if (regexp.charAt(i) == ']') {
+                if (!quoted && !escaped && insideList != 0) {
+                    insideList -= 1;
+                } else if (escaped) {
+                    escaped = false;
+                }
+            } else if (regexp.charAt(i) == 'Q') {
```

Comment:
I think it would be simpler to eliminate the `quoted` boolean and instead write a method that converts all \Q...\E into normal escaping.  That is slightly less efficient because it makes two passes over the string and might allocate a new string. but I think it will permit valuable simplification.  (I seem to recall that this is how the JDK's implementation handles \Q...\E.)

Reply:
We can do that or we can even eliminate the quoted part entirely because it doesn't contribute anything to the regex capturing groups-wise. Is there some side-effect of the latter?

Reply:
You are right that the quoted part is not needed.  So, having a little loop here that proceeds to the next \E (accounting for quoting) -- much like my suggestion for character classes -- makes a lot of sense and is better than the preprocessing that I suggested.